README.md
python/Depthcharge.md

CHANGELOG
2023-04-11: Depthcharge Release 0.5.0 ("Food for Thought")
----------------------------------------------------------------------------

* Added checker for CVE-2022-2347
* Added more helpful warning message resulting from a lack of a loadaddr


2022-07-05: Depthcharge Release 0.4.0 ("Entertainment!")
----------------------------------------------------------------------------

Added a handful of configuration checker builtins covering:

* CVE-2022-30552 and CVE-2022-30790 (#97)
* CVE-2021-27138 and CVE-2021-27097 (#99)
* No CVE: part_test_dos() stack overflow (#98)
* CVE-2020-8432
* CVE-2017-3225 and CVE-2017-3226
* CONFIG_ENV_IS_IN_* items (#90)
* CONFIG_FIT_SIGNATURE=y without CONFIG_FIT_FULL_CHECK=y


2022-03-05: Depthcharge Bug Fix Update 0.3.2
----------------------------------------------------------------------------

Fixed documentation rendering issues resulting from usage of the new 1.0.0
Sphinx RTD theme release and updated dependency specification accordingly.


2022-03-05: Depthcharge Bug Fix Update 0.3.1
----------------------------------------------------------------------------

Just a little releaselet to get a fix for [Issue #91] into main.

[Issue #91]: <https://github.com/nccgroup/depthcharge/issues/91>


2021-08-01: Depthcharge Release 0.3.0 ("Danny Nadelko")
----------------------------------------------------------------------------

Summary:

* Introduced AARCH64 support (#3)
* Added `Depthcharge.register_payload()` API function (#23)
* Added icache/dcache flush upon payload deployment (#24)
* Fixed ARM DataAbort parsing issue when in Thumb mode (#79)
* Fixed issue displaying multiple DRAM bank info (#81)
* Documentation updates and fixes
* Removal of RETURN_REGISTER payload.

Versions:

    Python module (`depthcharge`): 0.3.0
    Companion Firmware: 0.1.0

Details:

  **Introduced AARCH64 support**

  The depthcharge.arch.AARCH64 class has been introduced to support
  aarch64 system. A number of fixes and changes throughout the codebase 
  have been made to accommodate use on 64-bit systems.  

  There remains a known limitation on ARMv8 devices operating in AARCH32 bit
  mode, where we do not necessary have a generic means to trigger data aborts.

  The RETURN_REGISTER builtin payload and associated GoRegisterReader
  have been removed, as they ultimately provided little value, while being
  annoying to add AARCH64-bit support for. The RETURN_MEMORY payload (with no
  argument) can be used instead to return a gd pointer, if needed.  


2020-12-01: Depthcharge Release 0.2.0 ("Clatter and Hiss")
----------------------------------------------------------------------------

Summary:

* Introduced configuration checker API and depthcharge-audit-config
* Command-line argument and API "breaking" changes:
  * ARM is longer the default architecture, "Generic" is
  * Migrated uboot.py submodule to a uboot subpackage
  * Change to opt-in semantics for payload deploy+execute, device crash/reboot
  * Payload base and offset improvements

Versions:

    Python module (`depthcharge`): 0.2.0
    Companion Firmware: 0.1.0

Details:

  **Introduced configuration checker API and depthcharge-audit-config**

  The depthcharge.checker API is designed to to facilitate configuration
  auditing, where the goal is to consistently and quickly identify common
  security pitfalls or known vulnerabilities, so time spent on security
  review efforts can be focused elsewhere (e.g., in newly developed code).

  The depthcharge.checker.ConfigChecker class defines the overarching
  framework and implements the common auditing and reporting functionality.
  The depthcharge.checker.UBootConfigChecker and .UBootHeaderChecker
  classes implement support for parsing .config and platform configuration
  header files, respectively.

  The audit() methods of these "checkers" produce depthcharge.checker.Report
  objects, which are sets of depthcharge.checker.SecurityRisk objects
  that describe the potential risks and provide a high-level, general
  recommendations. The Report class supports a few export formats, including
  html, csv, and Markdown.

  The depthcharge-audit-config script exposes this new API functionality
  via the command-line.

  **ARM is longer the default architecture, "Generic" is.**

  In the interest of being more user friendly, Depthcharge now defaults to an
  ARM-ish "Generic" architecture, which is still 32-bit and little endian.

  By making no assumptions about registers and instruction set, we can exclude
  operations that would otherwise cause a target device to spin out of control.
  The goal here is try to balance provide some reasonable amount of
  functionality, while being resilient to platform-specific quirks that would
  otherwise cause grief fro a newcomer to Depthcharge.

  In order to use ARM-specific functionality, one must now do one of the
  following:

    * API: Create a Depthcharge context using `arch='arm'`
    * Scripts: pass an `--arch arm` argument

  See `arch/generic.py` for the other Generic architectures available.
  However, note that 64-bit and big-endian platforms have not yet undergone
  testing.

  **Migration uboot.py submodule to a uboot subpackage**

  The migration of the uboot.py submodule to a uboot subpackage was
  performed in order to allow new additions to be introduced in a more
  organized manner.

  The uboot.py submodule has been converted to a subpackage organized as
  follows:

     * depthcharge.uboot.board:      Platform and "board" configuration data
     * depthcharge.uboot.cmd_table:  Console command handler table inspection
     * depthcharge.uboot.env:        Environment (variable) functionality
     * depthcharge.uboot.jump_table: Exported jump table definitions

   This breaks the (unstable) v0.1.x API. Below is a summary of the v0.1.x
   items that have been removed, along with their corresponding replacements in
   the (also unstable) v0.2.x API.

   depthcharge.uboot.board:

python/depthcharge/arch.py
# SPDX-License-Identifier: BSD-3-Clause
# Depthcharge: <https://github.com/nccgroup/depthcharge>
#
"""
This module provides Architecture-specific information and helper functions.
"""

import os
import re
from copy import copy


class ArchitectureProperties(type):
    """
    Custom metaclass used to provide immutable class property accessors.
    """

    _name = None
    _word_size = None
    _phys_size = None
    _alignment = None
    _endianness = None
    _regs = None

    @property
    def name(cls) -> str:
        """
        Short, 1-word name for the architecture
        """
        return cls._name

    @property
    def description(cls) -> str:
        """
        Short description of the architecture
        """
        return cls._desc

    @property
    def alignment(cls) -> int:
        """
        Required word alignment, in bytes
        """
        return cls._alignment

    @property
    def word_size(cls) -> int:
        """
        Word size, in bytes.

        Currently, this class assumes ``word size == sizeof(int) == sizeof(void*)``,
        which is technically not consistent with the C standard.

        However, we'll cross that bridge when we get to it. (i.e., when introducing support for an
        architechture that breaks usages of :py:meth:`word_size`)

        If you're concerned about breaking changes, stick to the methods in
        this class rather than use word_size for directly.
        """
        return cls._word_size

    @property
    def phys_size(cls) -> int:
        """
        Size of phys_size_t, in bytes.
        """
        return cls._phys_size

    @property
    def endianness(cls) -> str:
        """
        Either ``'big'`` or ``'little'``

        Depthcharge does not currently support situations where an architecture supports both.
        (e.g. ARM in big-endian mode).
        """
        return cls._endianness

    @property
    def supports_64bit_data(cls) -> bool:
        """
        Indicates whether U-Boot has ``CONFIG_SYS_SUPPORT_64BIT_DATA`` for this
        architecture set.

        In general, this determines whether "qword" operations are supported by various operations
        and monitor commands.
        """
        return cls._supports_64bit_data

    @property
    def gd_register(cls) -> str:
        """
        This property provides name of the register that stores a pointer to U-Boot's global data
        structure (*gd_t*).

        A value of ``None`` is used to denote that the architecture does not store this infromation
        in a register.
        """
        for reg, info in cls._regs.items():
            if info.get('gd', False):
                return reg

        return None


class Architecture(metaclass=ArchitectureProperties):
    """
    Provides access to architecture-specific properties and conversion methods.
    """

    @classmethod
    def get(cls, arch_name):
        """
        Retrieve an architecture definition by name.

        **Example**:

        .. code:: python

            from depthcharge import Architecture

            arm = Architecture.get('arm')
            print(arm.description)
            for reg_name in arm.registers():
                print('  ' + reg_name)

        """
        for arch in cls.__subclasses__():
            if arch._name.lower() == arch_name.lower():
                return arch

        raise KeyError('No such architecture: ' + arch_name)

    @classmethod
    def supported(cls):
        """
        Returns a generator for all supported architectures.

        **Example**:

        .. code:: python

            from depthcharge import Architecture

            for arch in Architecture.supported():
                print(arch.name)

        """
        for arch in cls.__subclasses__():
            yield arch

    @classmethod
    def is_aligned(cls, address):
        """
        Returns ``True`` if *address* is word-aligned, and ``False`` otherwise.
        """
        return address & (cls._alignment - 1) == 0

    @classmethod
    def ptr_value(cls, data: bytes) -> int:
        """
        Read a pointer from the provided data and return this address as an
        integer in the host endianness.
        """
        return int.from_bytes(data[:cls._word_size], cls._endianness)

    @classmethod
    def ptr_value_adv(cls, data: bytes) -> int:
        """
        Returns a tuple *(value, data_slice)* such that the first element is the
        result of :py:meth:`ptr_value()`, and *data_slice* is *data* advanced by the size
        of the pointer.
        """
        return (cls.ptr_value(data), data[cls._word_size:])

    @classmethod
    def int_to_bytes(cls, intval: int) -> bytes:
        """
        Convert an integer value to `bytes`, according to the architecture's
        endianness and word size.
        """
        return intval.to_bytes(cls._word_size, cls._endianness)

    @classmethod
    def to_int(cls, data: bytes):
        """
        Returns the value in the provided data converted to a signed two's
        complement integer in the host endianness.
        """
        sign_bit = (1 << (cls._word_size * 8 - 1))
        mask     = (1 << (cls._word_size * 8)) - 1

        val = cls.to_uint(data)
        if val & sign_bit != 0:
            # Two's complement
            return -((~val & mask) + 1)

        return val

    @classmethod
    def to_int_adv(cls, data: bytes):
        """
        Returns a tuple *(value, data_slice)* such that the first element is the result of
        :py:meth:`to_int()`, and *data_slice* is *data* advanced by the size of the target's integer.
        """
        return (cls.to_int(data), data[cls._word_size:])

    @classmethod
    def to_uint(cls, data: bytes):
        """
        Returns the value in the provided data converted to an unsigned integer
        in the host endianness.
        """
        return int.from_bytes(data[:cls._word_size], cls._endianness)

    @classmethod
    def to_uint_adv(cls, data: bytes):
        """
        Returns a tuple *(value, data_slice)* such that the first element is the
        result of :py:meth:`to_uint()`, and *data_slice* is *data* advanced by the size of the
        target's (unsigned) integer.
        """
        return (cls.to_uint(data), data[cls._word_size:])

    @classmethod
    def hexint_to_bytes(cls, hex_str: str, num_bytes: int) -> int:
        """
        Convert a hexadecimal string representing an integer value to bytes,
        taking the architecture's endianness into account.
        """
        return int(hex_str, 16).to_bytes(num_bytes, cls._endianness)

    @classmethod
    def word_sizes(cls) -> dict:
        """
        This property provides a dictionary of word sizes supported by the
        architecture (int), mapped the command suffix used by U-Boot's memory
        commands (i.e., 'b', 'w', 'l', 'q').

        The quad-word suffix is only included if the architecture supports
        64-bit data, per :py:meth:`supports_64bit_data`.
        """
        ret = {1: 'b', 2: 'w', 4: 'l'}
        if cls.supports_64bit_data:
            ret[8] = 'q'
        return ret

    @classmethod
    def multiple_of_word_size(cls, value: int) -> bool:
        """
        Returns ``True`` if the provided *value* is a multiple of the
        architecture's word size, and ``False`` otherwise.
        """
        return value % cls._word_size == 0

    @classmethod
    def registers(cls) -> dict:
        """
        Dictionary containing registers names as keys and additional
        information (if any) in corresponding values.

        Special keys (i.e. not included with every entry) include:

            * `alias` - A string containing an alternate name for the register (e.g. `'sp'`).
            * `gd=True` if the register store's U-Boot global data structure.
            * `ident` - An integer identifier used internally (e.g. in payloads).

        """
        return copy(cls._regs)

    @classmethod
    def register(cls, name: str) -> tuple:
        """
        Look up a single register by case insensitve name or alias.

        The returned value is a tuple containing its corresponding key and value
        in the dictionary returned by :py:meth:registers()`.

        Raises :py:exc:`ValueError` for an invalid register name/alias.
        """
        for reg_name, info in cls._regs.items():
            if name in (reg_name, reg_name.upper(), reg_name.lower()):
                return (reg_name, info)

            try:
                alias = info['alias']
                if name in (alias, alias.upper(), alias.lower()):
                    return (reg_name, info)
            except KeyError:
                pass  # Not all registers have aliases

        raise ValueError('Invalid or unknown register: ' + name)


class ARM(Architecture):
    """
    ARMv7 (or earlier) target information - 32-bit little-endian
    """
    _name = 'ARM'
    _desc = 'ARM 32-bit, little-endian'
    _alignment = 4
    _word_size = 4
    _phys_size = 4
    _word_mask = 0xffffffff
    _endianness = 'little'
    _supports_64bit_data = False

    # ident values used by RETURN_REGISTER payload
    _regs = {
        'r0':  {'ident': 0x61},
        'r1':  {'ident': 0x62},
        'r2':  {'ident': 0x63},
        'r3':  {'ident': 0x64},
        'r4':  {'ident': 0x65},
        'r5':  {'ident': 0x66},
        'r6':  {'ident': 0x67},
        'r7':  {'ident': 0x68},
        'r8':  {'ident': 0x69},
        'r9':  {'ident': 0x6a, 'gd': True, 'alias': 'sb'},
        'r10': {'ident': 0x6b},
        'r11': {'ident': 0x6c, 'alias': 'fp'},
        'r12': {'ident': 0x6d, 'alias': 'ip'},
        'r13': {'ident': 0x6e, 'alias': 'sp'},
        'r14': {'ident': 0x6f, 'alias': 'lr'},
        'r15': {'ident': 0x70, 'alias': 'pc'},
    }

    _DA_ENTRY = re.compile(r"""
        (?P<name>[a-zA-Z][a-zA-Z0-9]+)
        \s?:\s?
        (?P<value>[0-9a-fA-F]{8})
    """, re.VERBOSE)

    @classmethod
    def parse_data_abort(cls, text: str) -> dict:
        """
        Parse ARM data abort output formatted as follows and return each field in a dict.

            00000001:data abort
            pc : [<8f7d8858>]	   lr : [<8f7d8801>]
            reloc pc : [<17835858>]	   lr : [<17835801>]
            sp : 8ed99718  ip : 00000000	 fp : 00000001
            r10: 00000001  r9 : 8eda2ea8	 r8 : 00000001
            r7 : 00000000  r6 : 00000004	 r5 : 00000004  r4 : 00000001
            r3 : 8ed9972c  r2 : 020200b4	 r1 : 8ed994ec  r0 : 00000009
            Flags: nZCv  IRQs off  FIQs off  Mode SVC_32
            Code: 2800f915 f04fd0cf e7ce30ff d10a2d04 (2000f8d8)

        """
        ret = {}
        for line in text.splitlines():
            pfx = ''
            if 'reloc' in line:
                pfx = 'reloc '
            elif line.startswith('Flags:'):
                ret['flags'] = {}
                for field in line.split('  '):
                    name, value = field.split(' ')
                    name = name.replace('Flags:', 'Asserted')
                    ret['flags'][name] = value
                continue

            elif line.startswith('Code:'):
                code = line.split()
                instructions = []
                for instruction in code[1:]:
                    try:
                        instruction = instruction.replace('(', '').replace(')', '').strip()
                        instruction = int(instruction, 16)
                        instruction = instruction.to_bytes(cls.word_size, byteorder=cls.endianness)
                        instructions.append(instruction)
                    except ValueError as e:
                        msg = 'Invalid instruction or parse error: ' + str(e)
                        raise ValueError(msg)

                ret['code'] = instructions

            else:
                for match in cls._DA_ENTRY.finditer(line):
                    regname, _ = cls.register(match.group('name'))
                    name = pfx + regname
                    value = match.group('value')

                    regs = ret.get('registers', {})

                    try:
                        regs[name] = int(value, 16)
                    except ValueError:
                        regs[name] = value

                    ret['registers'] = regs

        if not ret:
            msg = 'No data abort content found in the following text:' + os.linesep
            msg += text
            raise ValueError(msg)

        return ret

# TODO: class AArch64(Architecture)
# TODO: class MIPS(Architecture)
# TODO: class PPC(Architecture)
